src/core/lib/surface/call.cc

```diff
@@ -1740,8 +1794,1024 @@ void FilterStackCall::ContextSet(grpc_context_index elem, void* value,
+                gpr_log(GPR_DEBUG, "Append error: %s",
+                        absl::StrCat("key=", StringViewFromSlice(md->key),
+                                     " error=", error,
+                                     " value=", value.as_string_view())
+                            .c_str());
+              });
+  }
+}
+
+void PromiseBasedCall::ContextSet(grpc_context_index elem, void* value,
+                                  void (*destroy)(void*)) {
+  if (context_[elem].destroy) {
+    context_[elem].destroy(context_[elem].value);
+  }
+  context_[elem].value = value;
+  context_[elem].destroy = destroy;
+}
+
+void* PromiseBasedCall::ContextGet(grpc_context_index elem) const {
+  return context_[elem].value;
+}
+
+PromiseBasedCall::Completion PromiseBasedCall::StartCompletion(
+    void* tag, bool is_closure, const grpc_op* ops, size_t num_ops) {
+  Completion c(BatchSlotForOp(ops[0].op));
+  if (grpc_call_trace.enabled()) {
+    gpr_log(GPR_INFO, "%sStartCompletion %s tag=%p", DebugTag().c_str(),
+            c.ToString().c_str(), tag);
+  }
+  if (!is_closure) {
+    grpc_cq_begin_op(cq(), tag);
+  }
+  completion_info_[c.index()].pending = {
+      PauseReasonBit(PauseReason::kStartingBatch), is_closure, true, tag};
+  return c;
+}
+
+PromiseBasedCall::Completion PromiseBasedCall::PauseCompletion(
+    const Completion& completion, PauseReason reason) {
+  if (grpc_call_trace.enabled()) {
+    gpr_log(GPR_INFO, "%sPauseCompletion %s %s", DebugTag().c_str(),
+            completion.ToString().c_str(), PauseReasonString(reason));
+  }
+  auto& pause_bits = completion_info_[completion.index()].pending.pause_bits;
+  GPR_ASSERT((pause_bits & PauseReasonBit(reason)) == 0);
+  pause_bits |= PauseReasonBit(reason);
+  return Completion(completion.index());
+}
+
+void PromiseBasedCall::FailCompletion(const Completion& completion) {
+  if (grpc_call_trace.enabled()) {
+    gpr_log(GPR_INFO, "%sFailCompletion %s", DebugTag().c_str(),
+            completion.ToString().c_str());
+  }
+  completion_info_[completion.index()].pending.success = false;
+}
+
+void PromiseBasedCall::FinishCompletion(Completion* completion,
+                                        PauseReason reason) {
+  if (grpc_call_trace.enabled()) {
+    auto pause_bits = completion_info_[completion->index()].pending.pause_bits;
+    bool success = completion_info_[completion->index()].pending.success;
+    std::vector<const char*> pending;
+    for (size_t i = 0; i < 8 * sizeof(pause_bits); i++) {
+      if (static_cast<PauseReason>(i) == reason) continue;
+      if (pause_bits & (1 << i)) {
+        pending.push_back(PauseReasonString(static_cast<PauseReason>(i)));
+      }
+    }
+    gpr_log(GPR_INFO, "%sFinishCompletion %s %s %s", DebugTag().c_str(),
+            completion->ToString().c_str(), PauseReasonString(reason),
+            (pending.empty()
+                 ? (success ? std::string("done") : std::string("failed"))
```

Comment:
they're short strings so no allocation needed... i haven't timed it but i suspect it's probably better than strformat due to not needing to parse strings.

Reply:
To be clear, I wasn't proposing changing the way you're using `StrFormat()`; I was proposing calling `c_str()` on the result of `StrFormat()` rather than on the result of the `?:` operator.  In other words:

```
            (pending.empty()
                 ? (success ? "done" : "failed")
                 : absl::StrFormat("paused:remaining={%s}",
                                   absl::StrJoin(pending, ",")).c_str()));
```

Good point about the short strings not actually requiring any allocations.  But I think the code is probably a little more readable with this change anyway.

Up to you, though.